File: ObjectRequestBrokerCS/ObjectRequestBrokerCS/vson/parsers/ObjectParser.cs
using System;
using System.Collections.Generic;
using ObjectRequestBrokerCS.vson;
using Newtonsoft.Json.Linq;
using System.Reflection;

namespace ObjectRequestBrokerCS.vson.parsers
{

    public class ObjectParser
    {

        private static Dictionary<int, object> objectMap = null;

        public ObjectParser()
        {
        }


        public static object fromJson(string text)
        {
            JObject jsonObject = JObject.Parse(text);
            try
            {
                Type targetClass = Type.GetType(jsonObject.GetValue("type").ToString()); // get class name from "type" entry

                if (FieldUtils.isPrimitive(targetClass) || targetClass.Equals(typeof(string)))
                {
                    //ConstructorInfo targetConstructor = targetClass.GetConstructor(new[] { FieldUtils.attemptPrimitiveConvesion(targetClass) });
                    return targetClass.GetConstructor(new[] { FieldUtils.attemptPrimitiveConvesion(targetClass) }).Invoke(new[] { jsonObject.GetValue("value") });
                }


                ConstructorInfo targetConstructor = targetClass.GetConstructor(Type.EmptyTypes); // get noArgs constructor
                object targetObject = targetConstructor.Invoke(Type.EmptyTypes); // create an instance of targetClass
                JObject props = JObject.Parse(jsonObject.GetValue("properties").ToString());

                int hash = Convert.ToInt32(props.GetValue("@hash").ToString());

                if (objectMap.ContainsKey(hash))
                {
                    return objectMap[hash];
                }
                else
                {
                    FieldParser.fromJson(targetObject, props); // pass the instance to the field parser
                                                               // where all attribute key-val-pairs will be extracted from the "properties" field of the JSON
                    objectMap[hash] = targetObject;
                }
                return targetObject;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
                Console.Write(e.StackTrace);
                return null;
            }
        }

        public static void initObjectMap()
        {
            objectMap = new Dictionary<int, object>();
        }

        public static void clearObjectMap()
        {
            objectMap = null;
        }


    }
}

File: ObjectRequestBrokerJava/src/main/java/ORB/vson/parsers/ObjectParser.java
package ORB.vson.parsers;

import org.json.JSONObject;
import ORB.vson.FieldUtils;

import java.lang.reflect.Constructor;
import java.util.HashMap;

import static ORB.vson.FieldUtils.isPrimitive;

public class ObjectParser {

    private static HashMap<Integer, Object> objectMap = null;

    public ObjectParser() {
    }


    public static Object fromJson(String text) {
        JSONObject jsonObject = new JSONObject(text);
        try {
            Class<?> targetClass = Class.forName(jsonObject.get("type").toString().replace("System.","java.lang.")); // get class name from "type" entry

            if (isPrimitive(targetClass) || targetClass.equals(String.class)){
                Constructor targetConstructor = targetClass.getConstructor(FieldUtils.attemptPrimitiveConvesion(targetClass));
                Object targetObject = targetConstructor.newInstance(jsonObject.get("value"));
                return targetObject;
            }


            Constructor targetConstructor = targetClass.getConstructor(); // get noArgs constructor
            Object targetObject = targetConstructor.newInstance();  // create an instance of targetClass
            JSONObject props = new JSONObject(jsonObject.get("properties").toString());

            int hash = Integer.valueOf(props.get("@hash").toString());
            if (objectMap.containsKey(hash)) {
                return objectMap.get(hash);
            }
            else {
                FieldParser.fromJson(targetObject, props); // pass the instance to the field parser
                // where all attribute key-val-pairs will be extracted from the "properties" field of the JSON
                objectMap.put(hash,targetObject);
            }
            return targetObject;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static void initObjectMap() {
        objectMap = new HashMap<>();
    }

    public static void clearObjectMap() {
        objectMap = null;
    }


}
